rotkehlchen/chain/ethereum/modules/makerdao/sai/decoder.py

```diff
@@ -0,0 +1,521 @@
+            balance=Balance(),
+            location_label=cdp_creator,
+            counterparty=CPT_SAI,
+            notes=f'Create CDP {cdp_id}',
+        )
+        return event, []
+
+    def _decode_makerdao_close_cdp(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: list[HistoryBaseEntry],  # pylint: disable=unused-argument
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[list[ActionItem]],  # pylint: disable=unused-argument
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        cdp_creator = hex_or_bytes_to_address(tx_log.topics[1])
+        cdp_id = hex_or_bytes_to_int(tx_log.data[128:])
+
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.INFORMATIONAL and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.balance.amount == ZERO and
+                event.counterparty == CPT_SAI and
+                event.notes is not None and event.notes.startswith('Close CDP') is True
+            ):
+                return None, []
+
+        event = HistoryBaseEntry(
+            event_identifier=transaction.tx_hash,
+            timestamp=ts_sec_to_ms(transaction.timestamp),
+            sequence_index=tx_log.log_index,
+            location=Location.BLOCKCHAIN,
+            asset=self.eth,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.NONE,
+            balance=Balance(),
+            location_label=cdp_creator,
+            counterparty=CPT_SAI,
+            notes=f'Close CDP {cdp_id}',
+        )
+        return event, []
+
+    def _decode_makerdao_borrow_sai_event(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[list[ActionItem]],  # pylint: disable=unused-argument
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        cdp_id = hex_or_bytes_to_int(tx_log.topics[2])
+        withdrawer = hex_or_bytes_to_address(tx_log.topics[1])
+        amount_withdrawn_raw = hex_or_bytes_to_int(tx_log.topics[3])
+        amount_withdrawn = asset_normalized_value(amount=amount_withdrawn_raw, asset=self.sai)
+
+        already_decoded = False
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.GENERATE_DEBT and
+                event.balance.amount == amount_withdrawn and
+                event.counterparty == CPT_SAI
+            ):
+                already_decoded = True
+                return None, []
+
+        if already_decoded is False:
```

Comment:
Generally never do this. Do the opposite.
```suggestion
        if already_decoded is True:
            return None, []
```

But as I noticed you already `return None, []` up there ... so the `already_decoded` boolean is not even needed.

But my question stands. Why do you need this? The way our decoder works you should not get in this situation. Otherwise we would have needed this check in every single event decoding which means something is terribly broken

Reply:
The reason for this is to avoid duplication of events in a transaction e.g.

Sample txn: https://etherscan.io/tx/0x4aed2d2fe5712a5b65cb6866c51ae672a53e39fa25f343e4c6ebaa8eae21de80
<img width="958" alt="image" src="https://user-images.githubusercontent.com/72208758/207600246-35cb2684-978e-4e7a-8917-930191a62261.png">

SAI is only borrowed once but the event is emitted multiple times.

Reply:
But I removed the `already_decoded` variable as it was not needed.

Reply:
What I meant by duplication was that...

**Borrow 1,000 SAI From Maker** event occurs multiple times in the transaction whereas the user only receives the SAI once.

What I removed was the `already_decoded` variable as it was no longer needed. 
Why? Looping through `decoded_events` and returning early if an event as such already exists.

Is this a better explanation?

Reply:
Why does it occcur multiple times?